src/check_executor.rs

```diff
@@ -0,0 +1,169 @@
+use std::sync::Arc;
+
+use futures::StreamExt;
+use tokio::sync::mpsc::{self, UnboundedReceiver, UnboundedSender};
+use tokio::sync::oneshot::Sender;
+use tokio::task::JoinHandle;
+use tokio_stream::wrappers::UnboundedReceiverStream;
+
+use crate::checker::Checker;
+use crate::config_store::Tick;
+use crate::producer::ResultsProducer;
+use crate::types::check_config::CheckConfig;
+use crate::types::result::{CheckResult, CheckStatus, CheckStatusReasonType};
+
+#[derive(Debug)]
+pub struct ScheduledCheck {
+    pub tick: Tick,
+    pub config: Arc<CheckConfig>,
+    pub resolve_tx: Sender<CheckResult>,
+}
+
+pub fn run_executor(
+    concurrency: usize,
+    checker: Arc<impl Checker + 'static>,
+    producer: Arc<impl ResultsProducer + 'static>,
+) -> (UnboundedSender<ScheduledCheck>, JoinHandle<()>) {
+    tracing::info!("Starting check executor");
+
+    let (sender, reciever) = mpsc::unbounded_channel();
+    let executor =
+        tokio::spawn(async move { executor_loop(concurrency, checker, producer, reciever).await });
+
+    (sender, executor)
+}
+
+async fn executor_loop(
+    concurrency: usize,
+    checker: Arc<impl Checker + 'static>,
+    producer: Arc<impl ResultsProducer + 'static>,
+    reciever: UnboundedReceiver<ScheduledCheck>,
+) {
+    let schedule_check_stream: UnboundedReceiverStream<_> = reciever.into();
+
+    schedule_check_stream
+        .for_each_concurrent(concurrency, |scheduled_check| {
+            let job_checker = checker.clone();
+            let job_producer = producer.clone();
+
+            // TODO(epurkhiser): Record metrics on the size of the size of the
+
+            async move {
+                let ScheduledCheck {
+                    config,
+                    tick,
+                    resolve_tx,
+                } = scheduled_check;
+
+                let _ = tokio::spawn(async move {
+                    let check_result = job_checker.check_url(&config, &tick).await;
+
+                    if let Err(e) = job_producer.produce_checker_result(&check_result) {
+                        tracing::error!(error = ?e, "Failed to produce check result");
+                    }
```

Comment:
If we error here should we skip recording results and so on? I'm actually not sure what's best since the result failed to produce, so we kind of want to retry or block or something, since otherwise we're dropping results.